src/cryptsetup/cryptsetup-tpm2.c

```diff
@@ -64,7 +119,58 @@ int acquire_tpm2_key(
                 blob = loaded_blob;
         }
 
-        return tpm2_unseal(device, pcr_mask, pcr_bank, primary_alg, blob, blob_size, policy_hash, policy_hash_size, NULL, ret_decrypted_key, ret_decrypted_key_size);
+        if (flags & TPM2_FLAGS_USE_PIN) {
+                for (int i = 5;; i--) {
+                        _cleanup_(erase_and_freep) char *pin_str = NULL;
+
+                        if (i <= 0)
+                                return log_error_errno(SYNTHETIC_ERRNO(EPERM), "Failed to unlock, giving up!");
+
+                        r = get_pin(&pin_str);
+                        if (r == -EPERM)
+                                continue;
+                        if (r < 0)
+                                return r;
+
+                        r = tpm2_unseal(
+                                device,
+                                pcr_mask,
+                                pcr_bank,
+                                primary_alg,
+                                blob,
+                                blob_size,
+                                policy_hash,
+                                policy_hash_size,
+                                pin_str,
+                                ret_decrypted_key,
+                                ret_decrypted_key_size);
+                        /* We get this error in case there is an authentication policy mismatch. This should
+                         * not happen, but this avoids confusing behavior, just in case. */
+                        if (r == -EPERM)
+                                return r;
+                        if (r == -ENOLCK)
+                                return log_error_errno(
+                                        r,
+                                        "TPM2 is in dictionary attack lockout, no unlock possible right now.");
+                        if (r < 0)
+                                continue;
+
+                        return r;
```

Comment:
so this generates log messages on some errors, but not on others.

Usually our coding style says that functions should either log on all error conditions or on none (in which case the caller should do it). This function has some error paths that log and others which do not. 

See https://systemd.io/CODING_STYLE about this.

Reply:
OK, that makes sense (and avoids "double error messages"). The existing code isn't 100% consistent either but I'll try to make it no worse at least. :)

Reply:
yeah, there are some parts of the codebase we are really bad it (usually the older parts of the codebase). But we strive to getting better at it.